File: Stage28/Java/Problem3665.java
package Stage28.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Problem3665 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        int T = Integer.parseInt(br.readLine());

        while (T-- > 0) {
            int n = Integer.parseInt(br.readLine());

            int[] indegree = new int[n + 1];
            boolean[][] graph = new boolean[n + 1][n + 1];

            int[] rank = new int[n];

            st = new StringTokenizer(br.readLine());
            for (int i = 0; i < n; i++) rank[i] = Integer.parseInt(st.nextToken());

            for (int i = 0; i < n; i++) {
                for (int j = i + 1; j < n; j++) {
                    graph[rank[i]][rank[j]] = true;
                    indegree[rank[j]]++;
                }
            }

            int m = Integer.parseInt(br.readLine());

            while (m-- > 0) {
                st = new StringTokenizer(br.readLine());
                int a = Integer.parseInt(st.nextToken());
                int b = Integer.parseInt(st.nextToken());

                if (graph[a][b]) {
                    graph[a][b] = false;
                    graph[b][a] = true;
                    indegree[b]--;
                    indegree[a]++;
                } else {
                    graph[b][a] = false;
                    graph[a][b] = true;
                    indegree[a]--;
                    indegree[b]++;
                }
            }

            Queue<Integer> q = new LinkedList<>();

            for (int i = 1; i <= n; i++) {
                if (indegree[i] == 0) q.offer(i);
            }

            List<Integer> result = new ArrayList<>();
            boolean uncertain = false;

            for (int i = 0; i < n; i++) {
                if (q.isEmpty()) {
                    sb.append("IMPOSSIBLE\n");
                    uncertain = true;
                    break;
                }

                if (q.size() > 1) uncertain = true;

                int now = q.poll();
                result.add(now);

                for (int j = 1; j <= n; j++) {
                    if (graph[now][j]) {
                        indegree[j]--;
                        if (indegree[j] == 0) q.offer(j);
                    }
                }
            }

            if (!uncertain) {
                for (int x : result) sb.append(x).append(" ");
                sb.append("\n");
            } else if (!q.isEmpty()) {
                sb.append("?\n");
            }
        }

        System.out.println(sb);
        br.close();
    }
}

File: Stage28/C/Problem3665.c
#include <stdio.h>
#include <string.h>

int graph[501][501];
int indegree[501];
int queue[501];
int result[501];

int main(void) {
    int T;
    scanf("%d", &T);

    while (T--) {
        int n;
        scanf("%d", &n);

        memset(graph, 0, sizeof(graph));
        memset(indegree, 0, sizeof(indegree));

        int rank[501];

        for (int i = 0; i < n; i++)
            scanf("%d", &rank[i]);

        for (int i = 0; i < n; i++)
            for (int j = i + 1; j < n; j++) {
                graph[rank[i]][rank[j]] = 1;
                indegree[rank[j]]++;
            }

        int m;
        scanf("%d", &m);

        while (m--) {
            int a, b;
            scanf("%d %d", &a, &b);

            if (graph[a][b]) {
                graph[a][b] = 0;
                graph[b][a] = 1;
                indegree[b]--;
                indegree[a]++;
            } 
            else {
                graph[b][a] = 0;
                graph[a][b] = 1;
                indegree[a]--;
                indegree[b]++;
            }
        }

        int front = 0, rear = 0;

        for (int i = 1; i <= n; i++)
            if (indegree[i] == 0)
                queue[rear++] = i;

        int certain = 1;
        int cycle = 0;
        int idx = 0;

        for (int i = 0; i < n; i++) {

            if (front == rear) {
                cycle = 1;
                break;
            }

            if (rear - front > 1)
                certain = 0;

            int now = queue[front++];
            result[idx++] = now;

            for (int j = 1; j <= n; j++) {
                if (graph[now][j]) {
                    indegree[j]--;
                    if (indegree[j] == 0)
                        queue[rear++] = j;
                }
            }
        }

        if (cycle)
            printf("IMPOSSIBLE\n");

        else if (!certain)
            printf("?\n");

        else {
            for (int i = 0; i < n; i++)
                printf("%d ", result[i]);
            printf("\n");
        }
    }
}